### PR TITLE
Add TGLFBasedTransportModel

### PR DESCRIPTION
**Pull request by theo-brown**
Pulling out the good bits of #477 and bringing them up to date.

Aim for this PR is to focus on implementing the convertor between TORAX outputs and TGLF inputs.

### CODE REVIEW

**Comment by theo-brown** [comment]
Input conversion validated against JETTO TGLF runs for a spherical tokamak test case, using the EQDSK and IMAS equilibrium IDS from JETTO as the input to TORAX.
image(2)
image(1)

File: torax/_src/transport_model/tglf_based_transport_model.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+      geo: geometry.Geometry,
+      core_profiles: state.CoreProfiles,
+  ) -> TGLFInputs:
+    # TODO: Currently, geometry parameters are just taken from the global
+    # profiles. For complete accuracy, a Miller (or s-α) geometry fit routine
+    # should be implemented to calculate *local* geometry parameters.
+
+    # TODO: Currently, only a subset of TGLF inputs have been implemented.
+
+    # In this function, units and equations will be SI
```
Comment (Nush395) [inline torax/_src/transport_model/tglf_based_transport_model.py]: can this go in the method docstring?

File: torax/_src/transport_model/tglf_based_transport_model.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    # As a heads up, be warned TGLF uses a combination of:
+    # - SI with densities in 1e19 m^-3 and temperatures in keV
+    # - CGS Gaussian
+    # https://gafusion.github.io/doc/input_gacode.html
+    constants = constants_module.CONSTANTS
+    T_e = core_profiles.T_e.face_value() * constants.keV2J
+    n_e = core_profiles.n_e.face_value()
+
+    # Reference values used for TGLF-specific normalisation
```
Comment (Nush395) [inline torax/_src/transport_model/tglf_based_transport_model.py:186]: maybe similarly here this could go in the method docstring

File: torax/_src/transport_model/tglf_based_transport_model.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    q_prime = (
+        psi_calculations.calc_s_face(geo, core_profiles.psi)
+        * core_profiles.q_face**2
+        * a**2
+        / r**2
+    )
+
+    # Electron beta
+    # https://gafusion.github.io/doc/tglf/tglf_list.html#tglf-betae
+    # https://gafusion.github.io/doc/cgyro.html#faq
```
Comment (Nush395) [inline torax/_src/transport_model/tglf_based_transport_model.py]: the cgyro faq are also referred to above, could we have a small section on documentation that also has the main links in it for this model?

**Comment by theo-brown** [comment]
Thanks for taking a look at this @Nush395, I moved a bunch of stuff to the docstring as suggested.

File: torax/_src/transport_model/tglf_based_transport_model.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+@jax.tree_util.register_dataclass
+@dataclasses.dataclass(frozen=True)
+class TGLFInputs(quasilinear_transport_model.QuasilinearInputs):
+  r"""Dimensionless inputs to TGLF-based models.
+
+  See https://gafusion.github.io/doc/tglf/tglf_table.html for definitions.
+  """
+
+  # Ti/Te
+  Ti_over_Te: chex.Array
```
Comment (jcitrin) [inline torax/_src/transport_model/tglf_based_transport_model.py]: For all these typehints, should use the brand new array_typing.FloatVectorFace instead

https://github.com/google-deepmind/torax/blob/1c3c89e694e8e32c4e478cfad25f6f344e3074e9/torax/_src/array_typing.py#L35

File: torax/_src/transport_model/tglf_based_transport_model.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+# pylint: disable=invalid-name
+@jax.tree_util.register_dataclass
+@dataclasses.dataclass(frozen=True)
+class TGLFInputs(quasilinear_transport_model.QuasilinearInputs):
+  r"""Dimensionless inputs to TGLF-based models.
+
+  See https://gafusion.github.io/doc/tglf/tglf_table.html for definitions.
+  """
+
+  # Ti/Te
```
Comment (jcitrin) [inline torax/_src/transport_model/tglf_based_transport_model.py]: nit: cleaner to define all the dataclass attributes (no need to include the properties) in the dataclass docstring, instead of as comments

File: torax/_src/transport_model/tglf_based_transport_model.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+    # - 'r' in TGLF means the flux surface centroid minor radius. Gradients are
+    #   taken w.r.t. r
+    #   https://gafusion.github.io/doc/tglf/tglf_list.html#rmin-loc
+    # - B_unit = 1/r d(psi_tor/2π)/dr = q/2πr dpsi/dr, noting that psi_tor is
+    #   psi/2π
+    #   https://gafusion.github.io/doc/geometry.html#effective-field
+    #   https://gafusion.github.io/doc/cgyro.html#faq
+    # - c_s (ion sound speed)
+    #   https://gafusion.github.io/doc/cgyro/outputs.html#output-normalization
+    m_D_amu = 2.014  # Mass of deuterium [amu] TODO: load from lookup table
```
Comment (jcitrin) [inline torax/_src/transport_model/tglf_based_transport_model.py]: can do this TODO

File: torax/_src/transport_model/tglf_based_transport_model.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    # - TGLF uses a slightly different calculation of log_Lambda to those in
+    #   Wesson, namely
+    #       log_Lambda = (
+    #         24.0 - 0.5 * jnp.log(n_e_1e19 * 1e13) + jnp.log(T_e_keV * 1e3)
+    #       )
+    #   https://github.com/gafusion/gacode/blob/740bb2dc811bc1ad38be65ea4b87330995931305/cgyro/src/cgyro_make_profiles.F90#L145
+    #   This is different to Wesson by about ~0.5%. Below, we use the TLGF
+    #   version, but with n_e [m^-3] and T_e [J].
+    log_Lambda = 74.2 - 0.5 * jnp.log(n_e) + jnp.log(T_e)
+    nu_ee = (jnp.sqrt(2) * n_e * constants.qe**4 * log_Lambda) / (
```
Comment (jcitrin) [inline torax/_src/transport_model/tglf_based_transport_model.py]: avoid constants.qe**4 , this will underflow in fp32 . See how it's done here: https://github.com/google-deepmind/torax/blob/1c3c89e694e8e32c4e478cfad25f6f344e3074e9/torax/_src/physics/collisions.py#L240

**Comment by jcitrin** [comment]
FYI @gsnoep , @aaronkho

**Review by jcitrin — APPROVED** [review]

**Comment by theo-brown** [comment]
@jcitrin sorry I noticed some things that still needed doing. Is it ok for me to push to this branch now that you've marked it as `copybara:import-manual`, or is there a better option?

**Comment by jcitrin** [comment]
@theo-brown , sure go ahead and add. We hadn't actually started the internal review yet, no worries. Will hold off on it.

**Comment by theo-brown** [comment]
Thanks! I'm working on #1436 and this in parallel, having validated that the TGLFNNukaea implemented in fusion_surrogates is correctly reproducing the fluxes from the tglfnn-ukaea repo. Some issues I find are to do with how I'm talking to the NN, and some are with how I'm generating inputs/outputs, so there's a bit of bouncing between the two issues as I iterate. Hopefully should be able to get this in this week?

**Comment by theo-brown** [comment]
Failing tests are due to changes to `DynamicRuntimeParamsSliceProvider` on main, if I understand correctly.

**Comment by theo-brown** [comment]
Summary of changes since last review:
- Rebased onto main, removed DynamicRuntime objects
- Updated links to GACODE docs

**Comment by theo-brown** [comment]
@jcitrin this is ready for a re-review / for the copybara-import-manual label to be added back!

**Review by jcitrin — APPROVED** [review]
